src/test/java/test/listeners/github2415/SkipAfterPreviousTestClassFailureListener.java

```diff
@@ -0,0 +1,32 @@
+package test.listeners.github2415;
+
+import org.testng.IInvokedMethod;
+import org.testng.IInvokedMethodListener;
+import org.testng.ITestResult;
+import org.testng.SkipException;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class SkipAfterPreviousTestClassFailureListener implements IInvokedMethodListener {
+
+    private boolean hasFailures = false;
+    public static Map<String, Integer> status = new HashMap<>();
+
+    public void beforeInvocation(IInvokedMethod invokedMethod, ITestResult testResult) {
+        status.put(testResult.getTestClass().getName() + testResult.getName(), testResult.getStatus());
+        if (hasFailures) {
+            throw new SkipException("SkipException");
```

Comment:
I thought again about the issue and I think that the best way to do what you want is `testResult.setStatus(ITestResult.SKIP)` instead of `throw new SkipException("SkipException")`.
Please, could you add this additional test case in order to confirm it?

But supporting SkipException is a good feature too.

Once said that the fix should be catching SkipException and setting status to ITestResult.SKIP) somewhere.
I see the catch but I can't remember how the result could be SKIP (maybe a default value which can change a day).

@krmahadevan WDYT?

Reply:
@juherr I tried to set testResult as Skipped (testResult.setStatus(ITestResult.SKIP)) instead of throwing SkipException in beforeInvocation method. It doesn't stop test methods from second class even though there is testResult set as SKIP.

Reply:
I'd prefer that @krmahadevan confirm my opinion before but I think we have to support both.
And if we need to choose only one option, I prefer the second one because configuration methods are not supposed to throw an exception by definition (where the SkipException is a specific case).

Reply:
@juherr - Please give me sometime.. Currently swamped with stuff at work.. Will get to this maybe over this weekend.

Reply:
Ping @krmahadevan  :)

Reply:
@juherr my bad i missed this. Right now setting the ITestResult status as `SKIP` wouldnt have it skipped. That said and done, I think that would be almost equivalent to throwing the `SkipException` which kind of makes a lot more sense instead of setting the status for `ITestResult`. Within TestNG we should try and filter out the `SkipException` and handle it as a separate use case. 

On the whole this brings out a larger question.. Should listeners be allowed to throw exceptions, which will be handled by TestNG or should listeners be never allowed to throw exceptions (if they do, then the behavior would not be guaranteed), in which case, the only way for a person to cause a test to be skipped is by overriding its status.

Can we consider a third option which is to add a method into `ITestResult.forciblySkip()` which can be called by the listener implementation, and which can be inspected by TestNG and an appropriate action be taken ? That way we dont need to handle with exceptions etc., and we dont expect the user to fumble with the statuses as well.

Reply:
In fact, my question is more about the long term vision and not about the current implement. I'd like to determine the best approach for the long term vision.

As previously stated, an exception is only allowed in test methods where it will fail the test. Everywhere else, it is supposed to break the run.
The reason is that, in case of exception in a configuration method or in a listener, the user is the only one who knows what is the expected behavior between skip, fail, or anything else.

IMO, an exception should be exceptional and not used for flow control. That's why I don't really like `SkipException`.

There are 3 ways to skip a test method: 
* from a test method: `SkipException` is the more elegant approach if we don't want to too verbose but I think a test method is not the best place for the skip logic ;
* from a configuration method: `SkipException` can be used for verbose and coherence reasons but using `TestResult` should not be a big deal ;
* from a listener: `TestResult` is already provided and I don't see good reasons to support `SkipException` there.

In my opinion, the implementation of the SkipException should be as simple as:
```java
...
} catch (SkipException e) {
  result.setStatus(SKIPPED);
}
```
To be friendly user, the support in listener could be:
```java
try {
  listener. beforeInvocation(method, result);
} catch {
  Log.warning("SkipException catched from" + listener.getClass() + " : use result.setStatus(SKIPPED) instead of throwing SkipException");
  result.setStatus(SKIPPED);
}
```

For the record, how JUnit5 skip tests: https://nipafx.dev/junit-5-disabled-conditions/#disabled

@krmahadevan WDYT? Are we agree on the vision?
If we are, the next question will be to determine what is the first step to the goal :)

Reply:
@juherr - I still didnt get how are you proposing that a user be able to skip a test based on some condition from within a listener. 
That was where I was saying that perhaps we could introduce a new method within `ITestResult` interface called `skipInvocation()` which could be **ONLY** turned ON by a caller (cannot be turned off again) and which the user can flip ON via a listener.
That way, one could basically do the following:

1. Incorporate all the common logic based on which a test should be skipped within a listener ( putting this inside a test method causes it to be verbose and doing it from a configuration method causes the user to resort to inheritance if that logic is to be shared across multiple test classes). 
2. Listener invocation will now be free of this `SkipException` throwing as a special case to cause a test to be skipped.

IMO `SkipException` should only be thrown by tests and configurations (which I guess you are also saying the same and so we both are on the same page) and listeners should never throw exceptions. If they do, they will abort the run (which I guess is again what you are saying and so we are on the same page here as well)

Reply:
Ok, I understand now. Thanks for the clarification 👍 

`skipInvocation` is a better name than `forciblySkip` but the method is maybe useless on the API because I imagine the implementation as an equivalence of `setStatus(SKIP)`.
If we need `forciblySkip` for internal reasons, `setStatus` can be updated like:

```java
setStatus(status) {
  if (status == SKIP) {
    skipInvocation();
    return;
  }
  ...
}
```

Reply:
@juherr - Yeah.. `skipInvocation()` is only going to be equivalent to `setStatus(SKIP)` so it would just be another API. I was hoping that we could provide a clean intent to the user via this new method instead of having them to fiddle around with the status.